docs/src/doc/contribution/knowledge-base/memory-management.md

```diff
@@ -0,0 +1,32 @@
+
+### Memory Manager
+In C++ Godot, an object is 2 things, a base type and an optional script. We mirror this in the JVM with the wrappers and script instances. One major difference is that in C++, the script instance is simply a member of the base script and fake inheritance this way. In Kotlin, for technical reasons, we don't have another choice than doing true inheritance. A script instance always inherits a wrapper type. The consequences are that for a given C++ object, we can have 2 different JVM objects, each with their own lifecycle. The wrapper is bound to the Godot instance and the script instance is bound to its script. A Godot object can freely add, remove and switch script during its lifetime. In order to manage this, we have a singleton named `MemoryManager`. It got several functions:
+
+-  Keeping an up-to-date database of all Kotlin instances, being wrapper or script instance existing in the JVM.
+-  Matching any C++ pointer with an existing JVM instance (Necessary when a C++ parameter is sent to a Kotlin function)
+-  Synchronizing the life of JVM and C++ instances.
+
+![Memory Manager](../../assets/img/memory_manager.png)
+
+It's using a container simply named `GodotBinding`, holding a reference to both wrapper and script instance (when they exist). This is this type of object that is used as an `instance binding` back in C++. This means that as long as the C++ object is alive, the JVM instances will also stay alive.
+
+### Refcounted management
+`Refcounted` objects are a bit harder to manage as they don't rely on manual management but reference counting. Kotlin instances all contribute to increase that counter, so the native object won't die as long as the JVM needs it. It's a vicious cycle because throught the binding C++ is also holding a reference to the Kotlin instance, it means they are both keeping each other alive. If kept that way, both will never be able to be freed even if unused. The solution to this is to transform the instance binding into a weak reference when the C++ side no longer use it. That way, the Godot Binding will be able to be freed by the CG. When this happens, the MemoryManager will then decrement the counter of the C++ RefCounted and free it.
```

Comment:
```suggestion
`Refcounted` objects are a bit harder to manage as they don't rely on manual management but reference counting. Kotlin instances all contribute to increase that counter, so the native object won't die as long as the JVM needs it. It's a vicious cycle because throught the binding C++ is also holding a reference to the Kotlin instance, it means they are both keeping each other alive. If kept that way, both will never be able to be freed even if unused. The solution to this is to transform the instance binding into a weak reference when the C++ side no longer use it. That way, the Godot Binding will be able to be freed by the GC When this happens, the MemoryManager will then decrement the counter of the C++ RefCounted and free it.
```